orttraining/orttraining/core/optimizer/bias_dropout_fusion.cc

```diff
@@ -0,0 +1,200 @@
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    std::vector<std::reference_wrapper<Node>> nodes_to_fuse;
+
+    // matching for bias Add node
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) ||
+        node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    std::vector<NodeArg*> dropout_input, dropout_output;
+    const TensorShapeProto* input1_shape = node.MutableInputDefs()[0]->Shape();
+    const TensorShapeProto* input2_shape = node.MutableInputDefs()[1]->Shape();
+
+    if (input1_shape == nullptr ||
+        input2_shape == nullptr ||
+        input1_shape->dim_size() < 1 ||
+        input2_shape->dim_size() < 1) {
+      continue;
+    }
+
+    int last_dim_shape1 = input1_shape->dim_size() - 1;
+    int last_dim_shape2 = input2_shape->dim_size() - 1;
+    if (!utils::HasDimValue(input1_shape->dim(last_dim_shape1)) ||
+        !utils::HasDimValue(input2_shape->dim(last_dim_shape2)) ||
+        input1_shape->dim(last_dim_shape1).dim_value() != input2_shape->dim(last_dim_shape2).dim_value()) {
+      continue;
+    }
+
+    if (input1_shape->dim_size() == 1) {
+      dropout_input.push_back(node.MutableInputDefs()[1]);  // dropout input
+      dropout_input.push_back(node.MutableInputDefs()[0]);  // bias
+    } else if (input2_shape->dim_size() == 1) {
+      dropout_input.push_back(node.MutableInputDefs()[0]);  // dropout input
+      dropout_input.push_back(node.MutableInputDefs()[1]);  // bias
+    } else {
+      continue;
+    }
+    Node& add_node = node;
+    nodes_to_fuse.push_back(add_node);
+
+    // matching for Dropout node
+    auto next_node_itr = node.OutputNodesBegin();
+    if (next_node_itr == node.OutputNodesEnd()) {
+      continue;
+    }
+
+    const Node& next_node = (*next_node_itr);
+    if (!(graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "Dropout", {12}, kOnnxDomain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "TrainableDropout", {9}, kOnnxDomain)) ||
+        next_node.GetExecutionProviderType() != node.GetExecutionProviderType()) {
+      continue;
+    }
+
+    if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+      continue;
+    }
+
+    Node& dropout_node = *graph.GetNode(next_node.Index());
+    nodes_to_fuse.push_back(dropout_node);
+
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[0]);
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[1]);
+
+    FuseResidualAddIfAny(graph, dropout_node, dropout_input, dropout_output, nodes_to_fuse);
+
+    if (dropout_node.InputDefs().size() > 1) {
+      dropout_input.push_back(dropout_node.MutableInputDefs()[1]);  // ratio
+    }
+
+    // populate training_mode
+    bool is_trainable_dropout = (dropout_node.OpType() == "TrainableDropout");
```

Comment:
>is_trainable_dropout [](start = 9, length = 20)

Should we keep trainabledropout logic as it's gonna be deprecated soon? Or leave it here, and deal with it later when deprecated?

Reply:
Let's leave it here and remove it when it's deprecated.